### PR TITLE
[CBR-97] WIP! PR1: structured logging - introduce katip

### PR DESCRIPTION
**Pull request by CodiePP**
## Description

THIS IS WIP!  DO NOT NEVER EVER MERGE THIS!

* introduce `katip` library for logging
* provide compatible interface
* logging in IO with `Trace`

logging resides in util/Pos/Util/Log.hs
modules transformed to the new setup:
> 'util'
> 'networking'
> 'crypto'
> 'core'
> 'infra'
> 'txp'
> 'lrc'
> 'update'
> 'ssc'
> 'delegation'
> 'client'
> 'node-ipc'
> 'node'
> 'generator'

tests in these modules (if available), and benchmarks in 'networking', make use of the `katip` logging

## Linked issue

[CBR-97] U/S
[CBR-207] introduce katip
[CBR-271] logging configuration parser
[CBR-213] logging in IO - Pos.Util.Trace
[CBR-274] tests for Pos.Util.Log

## Type of change

- [~] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🛠 New feature (non-breaking change which adds functionality)
- [~] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [x] 🔨 New or improved tests for existing code
- [~] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [~] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## QA Steps

```
stack build cardano-sl-util
stack test cardano-sl-util
stack build cardano-sl-networking
stack test cardano-sl-networking
stack build --flag cardano-sl-networking:benchmarks cardano-sl-networking
stack build cardano-sl-infra
stack test cardano-sl-infra
stack build cardano-sl-txp
stack test cardano-sl-txp
stack build cardano-sl-lrc
stack test cardano-sl-lrc
stack build cardano-sl-update
stack build cardano-sl-ssc
stack build cardano-sl-delegation
stack build cardano-sl-client
stack test cardano-sl-client
stack build cardano-sl-generator
stack test cardano-sl-generator
stack build cardano-sl
stack test cardano-sl
stack build cardano-sl-node
stack build cardano-sl-node-ipc
```

### CODE REVIEW

**Comment by CodiePP** [comment]
This is WIP! Please, do not merge as modules other than 'util' and 'networking' may still fail to build.

File: util/Pos/Util/Log/Scribes.hs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+
+-------------------------------------------------------------------------------
+
+
+
+-- | global lock for stdout Scribe
+{-# NOINLINE lock #-}
+lock :: MVar ()
+lock = unsafePerformIO $ newMVar ()
```
[the remark this answers is not in the document]
Comment (avieth) [reply util/Pos/Util/Log/Scribes.hs]: I think it's OK to allow that, if two different scribes are made targeting the same `Handle`, there may be jumbled text due to concurrency.

**Comment by avieth** [comment]
I was once working on a branch to remove `Mockable`, but got a little carried away and also dumped log-warper into a little corner and replaced `WithLogger` with `Trace` parameters. I stopped because I don't want to step on your toes, but still you may be interested to see how I use it: https://github.com/input-output-hk/cardano-sl/tree/avieth/no_mockable/util/Pos/Util/Trace

That includes support for named logging, and for the "safe" logging (private vs. public log levels). It also deals with the various spots where a monad transformer is used to do "deffered" logging: log to a writer inside an STM transaction, then dump it all to log-warper when the transaction succeeds.

File: util/Pos/Util/Log/Scribes.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (CodiePP) [reply util/Pos/Util/Log/Scribes.hs]: yes, we had problems with concurrency and corrupted log entries when the lock was within the scribe providing function. Putting it one level up solved it.

File: util/Pos/Util/Log/Internal.hs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+{-# NOINLINE makeLSI #-}
+makeLSI :: MVar LoggingStateInternal
+-- note: only kick up tree if handled locally
+makeLSI = unsafePerformIO $ do
+    let lsiConfig = Nothing
+        lsiLogEnv = Nothing
+        lsiLinesLogged = 0
+    newMVar LoggingStateInternal {..}
```
[the remark this answers is not in the document]
Comment (CodiePP) [reply util/Pos/Util/Log/Internal.hs]: good point, I will try creating it in `setupLogging`

**Comment by erikd** [comment]
This is an astonishingly large PR! +3505 −1516 lines is huge! I don't even know where to begin reviewing this. On top of that there are merge conflicts to resolve and the git history is a mess.

If this PR was against Linux kernel and sent as a patch to the kernel people, it would simply be rejected as unmanagable. Can this not be split up into more easily digestible chunks?

File: explorer/src/explorer/Main.hs [hunk 3]
```diff
@@ -53,10 +54,12 @@ main = do
     args <- getExplorerNodeOptions
     let loggingParams = CLI.loggingParams loggerName (enaCommonNodeArgs args)
     loggerBracket loggingParams . logException "node" . runProduction $ do
+    --Log.loggerBracket Log.Info loggerName $ runProduction $ do
         logInfo "[Attention] Software is built with explorer part"
```
Comment (erikd) [inline explorer/src/explorer/Main.hs]: Shouldn't this commented out code be removed?

**Comment by kantp** [comment]
@CodiePP: I just had a quick chat with @erikd, about what can be done to make this PR easier to review and merge. I'm afraid this will be some work, but it's the last step needed to complete this huge effort.

One problem is that there is a mixture of pure whitespace changes (forced by the stylish haskell change) and semantic changes. In a small PR, this is not a big deal, but for such a large PR, any additional noise increases the difficulty of properly reviewing it. Untangling the whitespace and the non-trivial changes into separate commits would make the whole thing much easier to review.

The other thing is that due to the long-running nature of this branch, the history is not very helpful for the review process.

In order to achieve both, here's what we came up with:

- Start with the current `develop` branch, and run stylish-haskell on every file that is changed by this PR. Produce a separate PR to `develop` with those changes, and get it merged. This will be large, but easy to review, and it will contain all the semantically trivial whitespace changses 
- Afterwards, in principle, we could `rebase` the contents of this PR to the new `develop` branch. But this would be a _huge_ pain, since there would be intermittent merge conflicts for every commit. So instead, I suggest doing a pedestrian's rebase:
  - get all the changes from this PR into your working directory, by copying files, on top of the new `develop` including the whitespace changes
  - selectively stage changes (the emacs package `magit` is absolutely great for this) that belong together, to produce a reasonably sized, self-contained commit, and commit this.
  - repeat the step above until all the changes have ended up in commits
  - run a `git diff` on this branch and the one you would get if you merged this PR to the state of `develop` right before you started. If there are no other merges in between, this should be empty, otherwise, all the changes should belong to a PR that has been merged in between (for example, by merging in those branches and repeating the `git diff`, or by manual inspection if that should prove difficult). This step is to ensure that no changes got lost in the process
  - finally, open a new PR against `develop`. This one should be easier to review, since it only contains "real" code changes, and has a clean history.

It is quite some work, and very pedestrian, but I think it's the best way forward to finally get this reviewed and merged.

I would strongly suggest that we plan for a time window for this, and try to avoid any other merges to `develop` while this is ongoing.

**Comment by CodiePP** [comment]
Hi @kantp and @erikd, 
the first part is already done. We have merged with `develop` that contains the new 'stylish-haskell' changes, and we apply 'stylish-haskell' routinely to our changes. IMO, we cannot provide a separate PR for the changes of stylish-haskell because we have changed the imports in most (all?) of the files we touched. And, this is what stylish-haskell looks at.
I agree, that we can manually copy all changed files over to a fresh `develop`, branch off, and create a single commit and PR from this. This would compress the history of this feature branch.
For the time window: it is advisable to merge as much PRs as possible beforehand, just to cut down the number of open PRs. Any PR after our merging to `develop` will need to rebase, which leads to (possibly) many conflicting changes.

**Comment by vincenthz** [comment]
@erikd The linux kernel people would not have either applied stylistic changes to the overall code source nor did massive code churn in unexpected places. This really cut both ways.

otherwise yes, this PR is unreadable. simplest method to clean history here:
1. move to your topic branch then: `git reset --soft develop`, you should have all the diff in your checkout area. this is effectively what @kantp suggested, but instead using git.
2. commit all the diff (or alternatively for more work on your side, commit per hunk or line if you have a good way to make logical unit of commits)
3. force push and make a new PR.

**Comment by CodiePP** [comment]
we submit a new, clean PR

**Pull request closed without merging**